src/kernel.ts

```diff
@@ -325,7 +349,26 @@ class Kernel {
   }
 
   /**
-   * Send a message on the Kernel's shell channel.
+   * Disconnect the kernel.
+   */
+  disconnect(): void {
+    var inner_close = () => {
```

Comment:
I'd prefer to see this closure factored out into a private method. Something like:

```
disconnect(): void {
  if (this._ws !== null) {
    if (this._ws.readyState === WebSocket.OPEN) {
      this._ws.onclose = () => { this._clearSocket(); };
      this._ws.close();
    } else {
      this._clearSocket();
   }
}
```

This is just a style/readability thing. If you don't agree, that's fine and you can leave as-is.